clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -538,6 +548,116 @@ bindConstantVar = inlineBinders test
         _ -> return False
     -- test _ _ = return False
 
+-- | Push a cast over a case into it's alternatives.
+caseCast :: NormRewrite
+caseCast _ (Cast (Case subj ty alts) ty1 ty2) = do
+  alts' <- mapM castAlt alts
+  changed $ Case subj ty alts'
+    where
+      castAlt alt = do
+        (pat,altExpr) <- unbind alt
+        return $ bind pat (Cast altExpr ty1 ty2)
+caseCast _ e = return e
+
+-- | Push a cast over a Letrec into it's body
+letCast :: NormRewrite
+letCast _ (Cast (Letrec b) ty1 ty2) = do
+  let (binds,body) = unsafeUnbind b
+  changed $ Letrec $ bind binds (Cast body ty1 ty2)
+letCast _ e = return e
+
+
+-- | Push cast over an argument to a funtion into that function
+--
+-- This is done by specializing on the casted argument.
+-- Example:
+-- @
+--   y = f (cast a)
+--     where f x = g x
+-- @
+-- transforms to:
+-- @
+--   y = f' a
+--     where f' x' = (\x -> g x) (cast x')
+-- @
+argCastSpec :: NormRewrite
+argCastSpec ctx e@(App _ (Cast e' _ _)) = case e' of
+  Var _ _ -> go
+  Cast (Var _ _) _ _ -> go
+  _ -> warn go
+  where
+    go = specializeNorm ctx e
+    warn = trace (unlines ["WARNING: " ++ $(curLoc) ++ "specializing a function on a possibly non work-free cast."
+                          ,"Generated HDL implementation might contain duplicate work."
+                          ,"Please report this as a bug."
+                          ,""
+                          ,"Expression where this occurs:"
+                          ,showDoc e
+                          ])
+argCastSpec _ e = return e
+
+-- | Only inline casts that just contain a 'Var', because these are guaranteed work-free.
+-- These are the result of the 'splitCastWork' transformation.
+inlineCast :: NormRewrite
+inlineCast = inlineBinders test
+  where
+    test _ (_, Embed (Cast (Var _ _) _ _)) = return True
+    test _ _ = return False
+
+-- | Eliminate two back to back casts where the type going in and coming out are the same
+--
+-- @
+--   (cast :: b -> a) $ (cast :: a -> b) x   ==> x
+-- @
+eliminateCastCast :: NormRewrite
+eliminateCastCast _ c@(Cast (Cast e tyA tyB) tyB' tyC)
+  | tyB == tyB' = do
+    tcm <- Lens.view tcCache
+    let tyA' = normalizeType tcm tyA
+        tyC' = normalizeType tcm tyC
+    if tyA' == tyC' then changed e
+                    else throwError
+  | otherwise = throwError
+  where throwError = do
+          (nm,sp) <- Lens.use curFun
+          throw (ClashException sp ($(curLoc) ++ showDoc nm ++ ": Found 2 nested casts whose types don't line up:\n" ++ showDoc c) Nothing)
```

Comment:
Can we try to stay closer to a 80-column window in our formatting so that code is easier to read on smaller terminals?